### PR TITLE
ec2: Do not fallback to IMDSv1 on EC2

### PR DESCRIPTION
**Pull request by fred-lefebvre**
The EC2 Data Source needs to handle 3 states of the Instance
Metadata Service configured for a given instance :
1. HttpTokens : optional & HttpEndpoint : enabled
   Either IMDSv2 or IMDSv1 can be used.
2. HttpTokens : required & HttpEndpoint : enabled
   Calls to IMDS without a valid token (IMDSv1 or IMDSv2 with expired token)
   will return a 401 error.
3. HttpEndpoint : disabled
   The IMDS http endpoint will return a 403 error.

Previous work to support IMDSv2 in cloud-init handled case 1 and case 2. This commits handles case 3 by bypassing the retry block. It shaves 2 minutes when rebooting an instance that has had its IMDS http endpoint disabled but creates some inconsistencies.  An instance that doesn't set "manual_cache_clean" to "True" will have its ```/var/lib/cloud/instance``` symlink removed altogether after it has failed to find a datasource.

### CODE REVIEW

**Comment by fred-lefebvre** [comment]
Turning on ```manual_cache_clean``` seems a bit heavy-handed as it would lead to ignoring changes to block device mapping or attached ENIs. This is not a big deal for Amazon Linux but affect other distros that use cloud-init for network configuration.

Seems like an option that would enable the use the cached metadata only when a datasource can't be found would be useful. 
Any thoughts?

File: cloudinit/sources/DataSourceEc2.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def wait_for_metadata_service(self):
 
         # try the api token path first
         metadata_address = self._maybe_fetch_api_token(mdurls)
-        if not metadata_address:
+        if not metadata_address and self.cloud_name != CloudNames.AWS:
```
Comment (blackboxsw) [inline cloudinit/sources/DataSourceEc2.py]: This seems too broad a skip on CloudNames.AWS platform as we will now not fall back to IMDSv1 at all if we are AWS and there was a timeout on the IMDSv2 PUT for some reason.  No retries would be tried in this case and I believe the logic would take us down the path of not crawling any metadata in that case, also rejecting Ec2 as a datasource.
I really think we need to specifically check exc.code == 403 and ! AWS automatically fallback to IMDSv1 in that case.

File: cloudinit/sources/DataSourceEc2.py [hunk 2]
```diff
@@ -260,6 +260,8 @@ def wait_for_metadata_service(self):
         if metadata_address:
             self.metadata_address = metadata_address
             LOG.debug("Using metadata source: '%s'", self.metadata_address)
+        elif self.cloud_name == CloudNames.AWS:
```
Comment (blackboxsw) [inline cloudinit/sources/DataSourceEc2.py]: If we end up not finding a metadata_address in _wait_for_metadata_service then cloud-init will return an empty crawled_metadata dict in line 422 or def crawl_metadata. This would result in cloud-init saying that DataSourceEc2 is an invalid datasource for this vm and it would move on to other datasources. We want to always ensure we fallback to IMDSv1 even on ec2 lookalikes, if anything, this should b3 critical not a warning I'd say as I expect we'll not detect any datasource on this type of machine if we've gotten this far and still don't have a metadata_address

Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: Are there ec2 lookalikes for which ```CloudNames.AWS``` would be **True**?

To answer the 3 comments at once, I'm fine with checking specifically for the 403 error. Handling a timeout on the request for a new token is not a bad idea for falling back to IMDSv1 at that point is not desirable. On EC2, if the token request times out, we should retry ; going back to my question above about CloudNames.AWS and lookalikes.

I'll update the patch

Comment (blackboxsw) [reply cloudinit/sources/DataSourceEc2.py]: Thanks @fred-lefebvre for that, thanks for limiting the check to 403 errors.  I look forward to the changeset you mentioned.

To answer the question about Ec2-lookalikes, I'm not aware of any cloud platform the 'fakes'  being strictly CloudNames.AWS. The only way do do that would be for the platform to surface a DMI system-serial-number, chassis-asset-tag and system-serial-number to meet the following identitfication criteria: 

```
def identify_aws(data):
    # data is a dictionary returned by _collect_platform_data.
    if (data['uuid'].startswith('ec2') and
            (data['uuid_source'] == 'hypervisor' or
             data['uuid'] == data['serial'])):
        return CloudNames.AWS

    return None
```

If a platforms does fake DMI data to look like CloudNames.AWS, then they are beholden to feature sets and changes that true AWS surfaces, so they should be able to cope with IMDSv2 requirements. If they have special requirements otherwise that prevent them from running, then we'll need them to identify properly as a different ec2-lookalike CloudName.

So I think you don't have to worry about non-AWS ec2 lookalikes specifically here.
I was more concerned that AWS vms, on ec2-proper would  drop through during a timeout or unexpected error in  self._maybe_fetch_api_token() and skip straight to not detecting DataSourceEc2 at all if there were an issue with the IMDSv2 service being accessible during the vm boot and cloud-init datasource discovery.

Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: @blackboxsw : Thanks for the review.  Just to be clear, does  *"I look forward to the changeset you mentioned"* refer to my earlier comment about addressing the recovery when no data source can be discovered on an already configured instance?  I intend to take this on next but I would argue that this current change set can stand on its own and should be committed by itself to keep commits small and easier to review.

With the change here we improve IMDSv2 support by providing better logging  and a faster abort when the http endpoint is disabled.  Restricting the change to ```CloudNames.AWS``` makes it fairly safe.  Admittedly, the code here suffers from the same failings as the current cloud-init code base : If the IMDS endpoint is disabled, we are left without the ```/var/lib/cloud/instance``` symlink. But at least we get there faster. I would like to address the consistency issue in a separate PR as it will impact code outside of the EC2 datasource and will be easier to review on its own.  

**In short:** What do you want done before merging this PR?

**Comment by fred-lefebvre** [comment]
Moving this comment outside of the code review section for visibility:

@blackboxsw : Thanks for the review.  Just to be clear, does  *"I look forward to the changeset you mentioned"* refer to my earlier comment about addressing the recovery when no data source can be discovered on an already configured instance?  I intend to take this on next but I would argue that this current change set can stand on its own and should be committed by itself to keep commits small and easier to review.

With the change here we improve IMDSv2 support by providing better logging  and a faster abort when the http endpoint is disabled.  Restricting the change to ```CloudNames.AWS``` makes it fairly safe.  Admittedly, the code here suffers from the same failings as the current cloud-init code base : If the IMDS endpoint is disabled, we are left without the ```/var/lib/cloud/instance``` symlink. But at least we get there faster. I would like to address the consistency issue in a separate PR as it will impact code outside of the EC2 datasource and will be easier to review on its own.  

**In short:** What do you want done before merging this PR?

**Comment by blackboxsw** [comment]
@fred-lefebvre thanks for the response. I was going to say minimally I'd like to see unit tests covering this change and reducing the failure condition check to handling a 403.My plan is to get you a patch against this branch today for your review to see if that meets your needs.

How does that sound?

**Comment by fred-lefebvre** [comment]
> @fred-lefebvre thanks for the response. I was going to say minimally I'd like to see unit tests covering this change 

>and reducing the failure condition check to handling a 403.

Are you referring to this part?
```
       if isinstance(exception, uhelp.UrlError) and exception.code >= 400:
            if exception.code == 403:
                LOG.warning('Ec2 IMDS endpoint returned a 403 error. '
                            'HTTP endpoint is disabled. Aborting.')
            else:
                LOG.warning('Fatal error while requesting '
                            'Ec2 IMDSv2 API tokens')
            raise exception
```

The problem here is that there really is no recoverable error worth retrying when dealing with IMDS on EC2. ```401``` is the only recoverable error and it can't occur when requesting the token.  Anything else would indicate that we, somehow, malformed the request or that the endpoint itself has encountered an error... in which case, the best we can do is let the instance launch fail as fast as possible.
With the above code, we will only retry if we can't reach the IMDS endpoint on first try. Even then, 2 minutes seems to too long but it's consistent with existing code behavior.

>My plan is to get you a patch against this branch today for your review to see if that meets your needs.
> 
> How does that sound?

I agree about unit testing.  See my note above about the handling of the error codes.

File: cloudinit/sources/DataSourceEc2.py [hunk 3]
```diff
@@ -193,9 +192,19 @@ def get_instance_id(self):
...
+        On EC2. IMDS will always answer an API token, unless
+        the instance owner has disabled the IMDS HTTP endpoint or
+        the network topology conflics with the configured hop-limit.
+        """
         if self.cloud_name != CloudNames.AWS:
             return
 
+        url_params = self.get_url_params()
+        if url_params.max_wait_seconds <= 0:
+            return False
```
[the remark this answers is not in the document]
Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: I admit to copying this config validation from other calls to ```wait_for_urls```. I took it to mean I should not assume this the value of ```max_wait_seconds``` to be valid.

File: cloudinit/sources/DataSourceEc2.py [hunk 4]
```diff
@@ -496,20 +523,24 @@ def _refresh_stale_aws_token_cb(self, msg, exception):
...
-            self._api_token = API_TOKEN_DISABLED
+    def _imds_exception_cb(self, msg, exception=None):
+        if isinstance(exception, uhelp.UrlError) and exception.code >= 400:
+            if exception.code == 403:
+                LOG.warning('Ec2 IMDS endpoint returned a 403 error. '
+                            'HTTP endpoint is disabled. Aborting.')
+            else:
+                LOG.warning('Fatal error while requesting '
+                            'Ec2 IMDSv2 API tokens')
+            raise exception
```
[the remark this answers is not in the document]
Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: This is interesting data.  Can you get me instance IDs where you have seen this?  It doesn't match the guidance I have received  from folks maintaining IMDS. I would like to dive more into this.   I can modify the code to keep retrying on 400s for now and revisit later. Would that work for you?

File: cloudinit/sources/DataSourceEc2.py [hunk 5]
```diff
@@ -193,9 +192,19 @@ def get_instance_id(self):
             return self.metadata['instance-id']
 
     def _maybe_fetch_api_token(self, mdurls, timeout=None, max_wait=None):
+        """ Get an API token for EC2 Instance Metadata Service.
+
+        On EC2. IMDS will always answer an API token, unless
+        the instance owner has disabled the IMDS HTTP endpoint or
+        the network topology conflics with the configured hop-limit.
```
[the remark this answers is not in the document]
Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: will fix

File: cloudinit/sources/DataSourceEc2.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blackboxsw) [reply cloudinit/sources/DataSourceEc2.py]: @fred-lefebvre there error was on my part, well cloud-init's part in Focal only. 
Here's a fix that we will get into Ubuntu Focal which points to the details 
https://github.com/canonical/cloud-init/pull/230

LP:1865882

[the remark this answers is not in the document]
Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: > But I believe that was your desired outcome here right?

Yes. I'm removing the DataSourceNone from the list of data sources we use in our build (for now) to avoid the remapping of the instance_link to the fake instance id provided by DataSourceNone. This is actually the problem I'm trying to address here : https://github.com/canonical/cloud-init/pull/229

File: cloudinit/sources/DataSourceEc2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: fixed in latest rev.

File: cloudinit/sources/DataSourceEc2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (blackboxsw) [reply cloudinit/sources/DataSourceEc2.py]: In the case of Ec2, from docs,  it's actually how many seconds to wait  on a given url check before considering it as a non-viable source for metadata. In your case I don't think you want to skip the api token check if set to 0 because in that scenario people tried to reduce that timeout to make things faster, but on ec2 proper that would just break things fast without obtaining the token.

https://cloudinit.readthedocs.io/en/latest/topics/datasources/ec2.html#configuration

I think we should drop this part.

[the remark this answers is not in the document]
Comment (fred-lefebvre) [reply cloudinit/sources/DataSourceEc2.py]: done

**Comment by fred-lefebvre** [comment]
Updated with latest requested changes

**Comment by blackboxsw** [comment]
LP: https://bugs.launchpad.net/cloud-init/+bug/1866290
